### PR TITLE
Fix coverage workflow

### PR DESCRIPTION
**Pull request by kratman**
# Description

Coverage was not working properly, this should fix it

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: .github/workflows/test_on_push.yml [hunk 1]
```diff
@@ -106,13 +106,12 @@ jobs:
...
-        run: |
-          if [ ${{ matrix.python-version }} == "3.12" && ${{ matrix.os }} == "ubuntu-latest" ]; then
-            python -m nox -s coverage
-          else
-            python -m nox -s unit
-          fi
+        if: matrix.os != 'ubuntu-latest' || matrix.python-version != '3.12'
+        run: python -m nox -s unit
+
+      - name: Run unit/coverage tests for ${{ matrix.os }} with Python ${{ matrix.python-version }}
```
Comment (agriyakhetarpal) [inline .github/workflows/test_on_push.yml:112]:
```suggestion
      - name: Run coverage tests for ${{ matrix.os }} with Python ${{ matrix.python-version }}
```

**Review by agriyakhetarpal — COMMENTED** [review]
Suggestion: we should move the coverage upload step above the integration tests, since the coverage is just for the unit tests anyway – CodeCov will then analyse the coverage and return its results earlier

**Comment by agriyakhetarpal** [comment]
The notebook failure is sporadic and unrelated, should be fixed with a rerun

**Comment by agriyakhetarpal** [comment]
Another proposition: the `macos-14` images are much faster than the `ubuntu-latest` runners (~9 minutes vs ~25 minutes), so PyBOP had switched to using them for coverage – maybe we can try that as well? As far as I know, there is just one LaTeX test that needs to run on Linux, we can skip it on macOS to get faster coverage.

**Comment by kratman** [comment]
> Another proposition: the `macos-14` images are much faster than the `ubuntu-latest` runners (~9 minutes vs ~25 minutes), so PyBOP had switched to using them for coverage – maybe we can try that as well? As far as I know, there is just one LaTeX test that needs to run on Linux, we can skip it on macOS to get faster coverage.

For now, lets leave it. Especially if it requires skipping some coverage tests

**Comment by agriyakhetarpal** [comment]
The coverage upload failed silently again, maybe we need to supply GHA's special `${{ secrets.GITHUB_TOKEN }}` with it for it to work?

**Comment by kratman** [comment]
```
error - 2024-06-06 16:52:39,048 -- Upload failed: {"detail":"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 103 seconds."}
```

**Comment by agriyakhetarpal** [comment]
I think someone with admin permissions will have to add a secret to the repository, then – tagging @pybamm-team/maintainers

**Comment by kratman** [comment]
> I think someone with admin permissions will have to add a secret to the repository

https://github.com/codecov/codecov-action/pull/1471

I think it an issue with codecov because I am using a fork for my branch

**Comment by kratman** [comment]
Works now, sporadic failures will happen until the codecov fixes a bug

**Review by agriyakhetarpal — APPROVED** [review]

**Pull request merged**